src/codec.rs

```diff
@@ -0,0 +1,474 @@
+    }
+}
+
+/// Encode `items` into `bytes` as a variable-length opaque vector with a
+/// maximum length of `0xff`
+pub fn encode_items_opaque_u8<E: Encode>(bytes: &mut Vec<u8>, items: &[E]) {
+    let len_offset = bytes.len();
+    bytes.push(0);
+
+    for i in items {
+        i.encode(bytes);
+    }
+
+    let len = bytes.len() - len_offset - 1;
+    debug_assert!(len <= 0xff);
+    bytes[len_offset] = len as u8;
+}
+
+/// Decode `bytes` into a vector of `D` values, treating `bytes` as an opaque
+/// byte string of maximum length `0xff`.
+pub fn decode_items_opaque_u8<P, E: From<std::io::Error>, D: Decode<P, Error = E>>(
+    decoding_parameter: &P,
+    bytes: &mut Cursor<&[u8]>,
+) -> Result<Vec<D>, D::Error> {
+    // Read one byte to get length of opaque byte vector
+    let length = usize::from(u8::decode(&(), bytes)?);
+
+    decode_items_opaque(length, decoding_parameter, bytes)
+}
+
+/// Encode `items` into `bytes` as a variable-length opaque vector with a
+/// maximum length of `0xffff`
+pub fn encode_items_opaque_u16<E: Encode>(bytes: &mut Vec<u8>, items: &[E]) {
+    let len_offset = bytes.len();
+    bytes.extend(&[0, 0]);
+
+    for i in items {
+        i.encode(bytes);
+    }
+
+    let len = bytes.len() - len_offset - 2;
+    debug_assert!(len <= 0xffff);
+    let out: &mut [u8; 2] = (&mut bytes[len_offset..len_offset + 2]).try_into().unwrap();
+    *out = u16::to_be_bytes(len as u16);
+}
+
+/// Decode `bytes` into a vector of `D` values, treating `bytes` as an opaque
+/// byte string of maximum length `0xffff`.
+pub fn decode_items_opaque_u16<P, E: From<std::io::Error>, D: Decode<P, Error = E>>(
+    decoding_parameter: &P,
+    bytes: &mut Cursor<&[u8]>,
+) -> Result<Vec<D>, D::Error> {
+    // Read two bytes to get length of opaque byte vector
+    let length = usize::from(u16::decode(&(), bytes)?);
+
+    decode_items_opaque(length, decoding_parameter, bytes)
+}
+
+/// Decode the next `length` bytes from `bytes` into as many instances of `D` as
+/// possible.
+fn decode_items_opaque<P, E: From<std::io::Error>, D: Decode<P, Error = E>>(
+    length: usize,
+    decoding_parameter: &P,
+    bytes: &mut Cursor<&[u8]>,
+) -> Result<Vec<D>, D::Error> {
+    let mut decoded = Vec::new();
+    let initial_position = bytes.position() as usize;
+
+    // Create cursor over specified portion of provided cursor to ensure we
+    // can't read past len
+    let mut sub = Cursor::new(&bytes.get_ref()[initial_position..initial_position + length]);
+
+    while sub.position() < length as u64 {
+        decoded.push(D::decode(decoding_parameter, &mut sub)?);
+    }
+
+    // Advance outer cursor by the amount read in the inner cursor
+    bytes.set_position(initial_position as u64 + sub.position());
+
+    Ok(decoded)
+}
+
+/// Encode the `items` into `bytes` as a vector of `items.len()` items, up to
+/// `0xff`.
+pub fn encode_items_u8<E: Encode>(bytes: &mut Vec<u8>, items: &[E]) {
+    assert!(items.len() <= 0xff);
```

Comment:
Yes, you are right, and now that I think of it, it has to be this way: the point of the length prefix is so that TLS implementations that don't know about a particular message type can skip past it in a byte stream, and getting the number of elements doesn't help in that case. Thanks for catching that! This being the case, I think I can delete some of the `encode/decode_items_*` variants.